draft-davidson-pp-protocol.md

```diff
@@ -815,223 +589,105 @@ and is necessary for achieving the unforgeability requirement.
-  byte arrays into a single group element and output a new array
-  containing each of these elements.
-
-## API instantiation {#voprf-api}
+## Recommended ciphersuites {#voprf-ciph-recs}
 
-For the explicit signatures of each of the functions, refer to
-{{pp-functions}}.
+We recommend that the Issuer chooses one of the following ciphersuites
+detailed in {{I-D.irtf-cfrg-voprf}}:
 
-### ServerSetup
+- OPRF(curve448, SHA-512) (ID = 0x0002);
+- OPRF(P-384, SHA-512) (ID = 0x0004);
+- OPRF(P-521, SHA-512) (ID = 0x0005).
 
-~~~
-1. ciph = recover_ciphersuite_from_id(id)
-2. if ciph == null: panic(ERR_UNSUPPORTED_CONFIG)
-3. (k,Y,GG) = ciph.VerifiableSetup()
-4. key = k.as_bytes()
-5. pub_key = Y.as_bytes()
-6. s_cfg = ServerConfig {
-             s_id: id
-             ciphersuite: ciph,
-             key: key,
-             pub_key: pub_key,
-             max_evals: max_evals
-         }
-7. s_update = ServerUpdate {
-                s_id: id
-                ciphersuite: ciph,
-                pub_key: pub_key,
-                max_evals: max_evals
-            }
-8. Output (s_cfg, s_update)
-~~~
+We deliberately avoid the usage of smaller ciphersuites (associated with
+P-256 and curve25519) due to the potential to reduce security via
+static-DH attacks. See {{I-D.irtf-cfrg-voprf}} for more details.
 
-### ClientSetup
+## Protocol contexts
 
-~~~
-1. ciph = recover_ciphersuite_from_id(id)
-2. if ciph == null: panic(ERR_UNSUPPORTED_CONFIG)
-3. cfg = ClientConfig {
-            s_update: s_update
-         }
-4. Output cli_cfg
-~~~
+Note that we must run the verifiable version of the protocol in
+{{I-D.irtf-cfrg-voprf}}. Therefore the `Issuer` takes the role of the
+`Server` running in `modeVerifiable`. In other words, the `Issuer` runs
+`(ctxtI, pkI) = SetupVerifiableServer(suite)`; where `suite` is one of
+the ciphersuites in {{voprf-ciph-recs}}, `ctxt` contains the internal
+VOPRF Server functionality and secret key `skI`, and `pkI` is the Issuer
+public key. Likewise, run `ctxtC = SetupVerifiableClient(suite)` to
+generate the Client context.
+
+## Functionality {#voprf-api}
+
+For the explicit signatures of each of the functions, refer to
+{{pp-api}}.
 
 ### Generate
 
 ~~~
-1. ciph = cli_cfg.s.ciphersuite
-2. GG = ciph.group()
-3. c_data = []
-4. i_data = []
-5. g_data = []
-6. for i in 0..m:
-       1. c_data[i] = GG.scalar_field().random().as_bytes()
-7. (blinds,group_elems) = ciph.VerifiableBlind(c_data)
-8. for i in 0..m:
-       1. i_data[i] = group_elems[i].as_bytes()
-       2. g_data[i] = blinds[i].as_bytes()
-9. Output ClientIssuanceInput {
-              ClientIssuanceProcessing {
-                  client_data: c_data,
-                  gen_data: g_data,
-              },
-              ClientIssuanceElement {
-                  msg_data: i_data,
-              }
-          }
+def Generate(m):
+  inputs = []
+  for i in range(m):
+    inputs[i] = random_bytes()
+  (tokens, blindedTokens) = Blind(inputs)
```

Comment:
As we removed the batch interface to Blind, this pseudocode probably needs to be updated.